lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+"""A library to enable charms to implement "rolling" or "serialized" operations. E.g., a
+rolling restart.
+
+You may use this library directly, or extend it to customize behavior.
+
+For example, in order to implement a rolling restart, a charm author would need to add the
+following lines of code, to the following files (note the consistent use of the name "restart")
+to namespace this particular rolling op:
+
+Add a peer relation to `metadata.yaml`:
+```yaml
+peers:
+    restart:
+        interface: rolling_op
+```
+
+To enable a human operator to trigger the restat, add an action to `actions.yaml`:
+```yaml
+restart:
+  description: Restarts a service
+  params: {}
+```
+
+Import the library, and enable it by doing the following:
+
+1. Override the Charm's `on` property with the RollingEvents class from the library.
+2. Add a method containing code to be executed to the Charm object. Here, we call this
+method `_restart`.
+3. Initialize a RollingOpsManager class, passing in the Charm, name of the peer relation,
+restart handler, and action.
+
+src/charm.py
+```python
+# ...
+from charms.rolling_ops.v0.rollingops import RollingOpsManager, RollingEvents
+# ...
+
+class SomeCharm(...):
+    # ...
+    on = RollingEvents
+    # ...
+
+    def __init__(...)
+        # ...
+        self.restart = RollingOpsManager(self, 'restart', self._restart, self.on.restart_action)
+        # ...
+
+    def _restart(self, event):
+        systemd.service_restart('foo')
+```
+
+"""
+import logging
+from ops.charm import CharmBase, RelationChangedEvent, CharmEvents, ActionEvent
+from ops.framework import EventBase, EventSource, Object
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "20b7777f58fe421e9a223aefc2b4d3a4"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class LockErrorNoRelation(Exception):
+    pass
```

Comment:
This could do with a docstring explaining when/why this exception would be raised :)

Reply:
It gets raised when there is no relation. But yes, I can add a docstring :-)